File: prob1.java
// Time Complexity : O(logn) 
// Space Complexity : O(logn)
// Did this code successfully run on Leetcode : Yes

class Solution {
    int[] start = new int[] {0, 1, 6, 8, 9};

    public int confusingNumberII(int n) {
        return count(n, 1) + count(n, 6) + count(n, 8) + count(n, 9);
    }

    int count(int n, long current) {
        
        if(current > n) return 0;
        var flipped = flip(current);
        int count = 0;
        // we can count the value if they are not the same. so we dont  count numbers like 1, 8, 88 etc..
        if(flipped != current) count++;
        for(var s: start) {
// we just take the current number and append the next digit in the least significant location. ex 10 becomes 100, 101, 106, 108 and 109
            var next = current * 10 + s;
            count += count(n, next);
        }
        return count;
    }

    long flip(long num) {
    // we use this method to flip a number. so 1998 becomes 8661
        long result = 0;
        while(num > 0) {
            var digit =(int) num % 10;
            num/=10;
            result*=10;
            result+=flippedDigit(digit);
        }
        return result;
    }

    int flippedDigit(int digit) {
        // just helps us flip a single digit
        if(digit == 9) return 6;
        if(digit == 6) return 9;
        return digit;
    }
}

File: Prob1.py
class Solution:
    def confusingNumberII(self, n: int) -> int:
        #Method 1 - Check if each number within range is confusing -> Brute force

        #Method 2 - DFS by geenrating numbers from 0,1,6,8,9 instead of checking each number in the given range.
        #Since, we know only 0,1,6,8,9 can cause cosufing numebrs -> map them with thier rotates.
        confusing_pairs = {0: 0, 1: 1, 6: 9, 8: 8, 9: 6}
        count=0
        def isconfusingnum(num):
            temp=num
            res=0
            while num>0:
                digit=num%10
                res=res*10+confusing_pairs[digit]
                num=num//10
            return res!=temp #if result is same as current num-> false else True

        def dfs(currnum,n):
            nonlocal count
            #base
            if currnum>n: return #when we go out of range->return

            #logic
            if isconfusingnum(currnum): 
                count+=1
            for k in confusing_pairs.keys(): #use each key (0,1,6,8,9) and generate the new number in the next level of the tree and dfs.
                newnum=currnum*10+k
                if newnum != 0: dfs(newnum, n)        
        dfs(0,n)
        return count
